[TypeChecker] Fix crash with Objective C keypaths.

rudkx opened the pull request:
We have a crash reported but do not have a test case and my attempt at
creating one was unsuccessful. It seems like the issue is that we have
an invalid decl that we attempt to access the interface type on,
though, so hopefully this fixes the issue.

I also tweaked the constraint generator to bail out when
checkObjCKeyPathExpr fails.

Potentially fixes: rdar://problem/40955755

Commit:
[TypeChecker] Fix crash with Objective C keypaths.

We have a crash reported but do not have a test case and my attempt at
creating one was unsuccessful. It seems like the issue is that we have
an invalid decl that we attempt to access the interface type on,
though, so hopefully this fixes the issue.

I also tweaked the constraint generator to bail out when
checkObjCKeyPathExpr fails.

Potentially fixes: rdar://problem/40955755

rudkx commented:
@swift-ci Please smoke test

slavapestov commented on lib/Sema/CSGen.cpp:
```diff
@@ -3280,7 +3280,8 @@ namespace {
         if (keyPath->isObjC()) {
           auto &cs = CG.getConstraintSystem();
-          (void)cs.getTypeChecker().checkObjCKeyPathExpr(cs.DC, keyPath);
+          if (!cs.getTypeChecker().checkObjCKeyPathExpr(cs.DC, keyPath))
```
Do we emit a diagnostic when this returns false?

rudkx replied:
Yes, at least in the cases we already hit this in during test runs.

slavapestov commented on lib/Sema/TypeCheckExprObjC.cpp:
```diff
@@ -317,6 +317,8 @@ Optional<Type> TypeChecker::checkObjCKeyPathExpr(DeclContext *dc,
     // Handle property references.
     if (auto var = dyn_cast<VarDecl>(found)) {
       validateDecl(var);
+      if (var->isInvalid() || !var->hasInterfaceType())
```
Try a property with no declared type and an initial value expression that references itself via a keypath

rudkx replied:
To be clear, I can get this code to be hit, but not with a nullptr interface type. Instead, we have an error type.

I'll push another commit that checks explicitly for error types and adds a test case that exercises that path.

It would be nice to understand how this was actually reaching this point with a null type, though - it's probably indicative of some bad path in decl validation.

Commit:
Also check for error_type in ObjC keypath validation.

And add a test that exercises this exit path (although it won't crash
a compiler without my fixes because we do end up with ErrorType in
that case, not a nullptr).

rudkx commented:
@swift-ci Please smoke test and merge

rudkx commented:
@swift-ci Please smoke test
